runtime/android/core/src/org/xwalk/core/XWalkActivityDelegate.java

```diff
@@ -26,16 +32,57 @@
     private boolean mIsXWalkReady;
     private boolean mBackgroundDecorated;
     private boolean mWillDecompress;
+    private final boolean mIsDownloadMode;
 
     public XWalkActivityDelegate(Activity activity,
             Runnable cancelCommand, Runnable completeCommand) {
         mActivity = activity;
         mCancelCommand = cancelCommand;
         mCompleteCommand = completeCommand;
-
-        mDialogManager = new XWalkDialogManager(mActivity);
-        mXWalkUpdater = new XWalkUpdater(this, mActivity, mDialogManager);
-
+        mIsDownloadMode = isDownloadMode();
+
+        if (mIsDownloadMode) {
+            mXWalkUpdater = new XWalkUpdater(
+                new XWalkBackgroundUpdateListener() {
+                    @Override
+                    public void onXWalkUpdateStarted() {
+                    }
+
+                    @Override
+                    public void onXWalkUpdateProgress(int percentage) {
+                    }
+
+                    @Override
+                    public void onXWalkUpdateCancelled() {
+                        mCancelCommand.run();
+                    }
+
+                    @Override
+                    public void onXWalkUpdateFailed() {
+                        Toast.makeText(mActivity,
+                                "Crosswalk Runtime Update Failed!",
+                                Toast.LENGTH_SHORT).show();
+                        mCancelCommand.run();
```

Comment:
As mentioned before, showing toast breaks the basic principle "silent" of silent download mode. And we can't just finish the activity when download failed. In shared mode, there's a dialog interacting with the end-user, and all other interactions are interrupted. The cancel command will only be called when the end-user selected to cancel. But in download mode, all things are done in background and the end-user may be input something in foreground. We can't just send a notification and then close the activity without approval. The best solution is letting the developer to handle this kind of situation.